File: java/src/main/java/mal/reader.java
package mal;

import java.util.ArrayList;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import org.apache.commons.lang3.StringEscapeUtils;
import mal.types.*;

public class reader {
    public static class ParseError extends MalThrowable {
        public ParseError(String msg) {
            super(msg);
        }
    }

    public static class Reader {
        ArrayList<String> tokens;
        Integer position;
        public Reader(ArrayList<String> t) {
            tokens = t;
            position = 0;
        }

        public String peek() {
            if (position >= tokens.size()) {
                return null;
            } else {
                return tokens.get(position);
            }
        }
        public String next() {
            return tokens.get(position++);
        }
    }

    public static ArrayList<String> tokenize(String str) {
        ArrayList<String> tokens = new ArrayList<String>();
        Pattern pattern = Pattern.compile("[\\s ,]*(~@|[\\[\\]{}()'`~@]|\"(?:[\\\\].|[^\\\\\"])*\"|;.*|[^\\s \\[\\]{}()'\"`~@,;]*)");
        Matcher matcher = pattern.matcher(str);
        while (matcher.find()) {
            String token = matcher.group(1);
            if (token != null &&
                !token.equals("") &&
                !(token.charAt(0) == ';')) {
                tokens.add(token);
            }
        }
        return tokens;
    }

    public static MalVal read_atom(Reader rdr)
            throws ParseError {
        String token = rdr.next();
        Pattern pattern = Pattern.compile("(^-?[0-9]+$)|(^-?[0-9][0-9.]*$)|(^nil$)|(^true$)|(^false$)|^\"(.*)\"$|:(.*)|(^[^\"]*$)");
        Matcher matcher = pattern.matcher(token);
        if (!matcher.find()) {
            throw new ParseError("unrecognized token '" + token + "'");
        }
        if (matcher.group(1) != null) {
            return new MalInteger(Integer.parseInt(matcher.group(1)));
        } else if (matcher.group(3) != null) {
            return types.Nil;
        } else if (matcher.group(4) != null) {
            return types.True;
        } else if (matcher.group(5) != null) {
            return types.False;
        } else if (matcher.group(6) != null) {
            return new MalString(StringEscapeUtils.unescapeJson(matcher.group(6)));
        } else if (matcher.group(7) != null) {
            return new MalString("\u029e" + matcher.group(7));
        } else if (matcher.group(8) != null) {
            return new MalSymbol(matcher.group(8));
        } else {
            throw new ParseError("unrecognized '" + matcher.group(0) + "'");
        }
    }

    public static MalVal read_list(Reader rdr, MalList lst, char start, char end)
            throws MalContinue, ParseError {
        String token = rdr.next();
        if (token.charAt(0) != start) {
            throw new ParseError("expected '" + start + "'");
        }

        while ((token = rdr.peek()) != null && token.charAt(0) != end) {
            lst.conj_BANG(read_form(rdr));
        }

        if (token == null) {
            throw new ParseError("expected '" + end + "', got EOF");
        }
        rdr.next();

        return lst;
    }

    public static MalVal read_hash_map(Reader rdr)
            throws MalContinue, ParseError {
        MalList lst = (MalList)read_list(rdr, new MalList(), '{', '}');
        return new MalHashMap(lst);
    }

    public static MalVal read_form(Reader rdr)
            throws MalContinue, ParseError {
        String token = rdr.peek();
        if (token == null) { throw new MalContinue(); }
        MalVal form;

        switch (token.charAt(0)) {
            case '\'': rdr.next();
                       return new MalList(new MalSymbol("quote"),
                                          read_form(rdr));
            case '`': rdr.next();
                      return new MalList(new MalSymbol("quasiquote"),
                                         read_form(rdr));
            case '~':
                if (token.equals("~")) {
                    rdr.next();
                    return new MalList(new MalSymbol("unquote"),
                                       read_form(rdr));
                } else {
                    rdr.next();
                    return new MalList(new MalSymbol("splice-unquote"),
                                       read_form(rdr));
                }
            case '^': rdr.next();
                      MalVal meta = read_form(rdr);
                      return new MalList(new MalSymbol("with-meta"),
                                         read_form(rdr),
                                         meta);
            case '@': rdr.next();
                      return new MalList(new MalSymbol("deref"),
                                         read_form(rdr));
            case '(': form = read_list(rdr, new MalList(), '(' , ')'); break;
            case ')': throw new ParseError("unexpected ')'");
            case '[': form = read_list(rdr, new MalVector(), '[' , ']'); break;
            case ']': throw new ParseError("unexpected ']'");
            case '{': form = read_hash_map(rdr); break;
            case '}': throw new ParseError("unexpected '}'");
            default:  form = read_atom(rdr);
        }
        return form;
    }

    public static MalVal read_str(String str)
            throws MalContinue, ParseError {
        return read_form(new Reader(tokenize(str)));
    }
}

File: kotlin/src/mal/reader.kt
package mal

import kotlin.text.Regex

val TOKEN_REGEX = Regex("[\\s,]*(~@|[\\[\\]{}()'`~^@]|\"(?:\\\\.|[^\\\\\"])*\"|;.*|[^\\s\\[\\]{}('\"`,;)]*)")
val ATOM_REGEX = Regex("(^-?[0-9]+$)|(^nil$)|(^true$)|(^false$)|^\"(.*)\"$|:(.*)|(^[^\"]*$)")

class Reader(sequence: Sequence<String>) {
    val tokens = sequence.iterator()
    var current = advance()

    fun next(): String? {
        var result = current
        current = advance()
        return result
    }

    fun peek(): String? = current

    private fun advance(): String? = if (tokens.hasNext()) tokens.next() else null
}

fun read_str(input: String?): MalType {
    val tokens = tokenizer(input) ?: return NIL
    return read_form(Reader(tokens))
}

fun tokenizer(input: String?): Sequence<String>? {
    if (input == null) return null

    return TOKEN_REGEX.findAll(input)
            .map({ it -> it.groups[1]?.value as String })
            .filter({ it != "" && !it.startsWith(";")})
}

fun read_form(reader: Reader): MalType =
        when (reader.peek()) {
            null -> throw MalContinue()
            "("  -> read_list(reader)
            ")"  -> throw MalReaderException("expected form, got ')'")
            "["  -> read_vector(reader)
            "]"  -> throw MalReaderException("expected form, got ']'")
            "{"  -> read_hashmap(reader)
            "}"  -> throw MalReaderException("expected form, got '}'")
            "'"  -> read_shorthand(reader, "quote")
            "`"  -> read_shorthand(reader, "quasiquote")
            "~"  -> read_shorthand(reader, "unquote")
            "~@" -> read_shorthand(reader, "splice-unquote")
            "^"  -> read_with_meta(reader)
            "@"  -> read_shorthand(reader, "deref")
            else -> read_atom(reader)
        }

fun read_list(reader: Reader): MalType = read_sequence(reader, MalList(), ")")
fun read_vector(reader: Reader): MalType = read_sequence(reader, MalVector(), "]")

private fun read_sequence(reader: Reader, sequence: IMutableSeq, end: String): MalType {
    reader.next()

    do {
        val form = when (reader.peek()) {
            null -> throw MalReaderException("expected '$end', got EOF")
            end  -> { reader.next(); null }
            else -> read_form(reader)
        }

        if (form != null) {
            sequence.conj_BANG(form)
        }
    } while (form != null)

    return sequence
}

fun read_hashmap(reader: Reader): MalType {
    reader.next()
    val hashMap = MalHashMap()

    do {
        var value : MalType? = null;
        val key = when (reader.peek()) {
            null -> throw MalReaderException("expected '}', got EOF")
            "}"  -> { reader.next(); null }
            else -> {
                var key = read_form(reader)
                if (key !is MalString) {
                    throw MalReaderException("hash-map keys must be strings or keywords")
                }
                value = when (reader.peek()) {
                    null -> throw MalReaderException("expected form, got EOF")
                    else -> read_form(reader)
                }
                key
            }
        }

        if (key != null) {
            hashMap.assoc_BANG(key, value as MalType)
        }
    } while (key != null)

    return hashMap
}

fun read_shorthand(reader: Reader, symbol: String): MalType {
    reader.next()

    val list = MalList()
    list.conj_BANG(MalSymbol(symbol))
    list.conj_BANG(read_form(reader))

    return list
}

fun read_with_meta(reader: Reader): MalType {
    reader.next()

    val meta = read_form(reader)
    val obj = read_form(reader)

    val list = MalList()
    list.conj_BANG(MalSymbol("with-meta"))
    list.conj_BANG(obj)
    list.conj_BANG(meta)

    return list
}

fun read_atom(reader: Reader): MalType {
    val next = reader.next() ?: throw MalReaderException("Unexpected null token")
    val groups = ATOM_REGEX.find(next)?.groups ?: throw MalReaderException("Unrecognized token: " + next)

    return if (groups[1]?.value != null) {
        MalInteger(groups[1]?.value?.toLong() ?: throw MalReaderException("Error parsing number: " + next))
    } else if (groups[2]?.value != null) {
        NIL
    } else if (groups[3]?.value != null) {
        TRUE
    } else if (groups[4]?.value != null) {
        FALSE
    } else if (groups[5]?.value != null) {
        MalString((groups[5]?.value as String).replace("\\n", "\n").replace("\\\"", "\"").replace("\\\\", "\\"))
    } else if (groups[6]?.value != null) {
        MalKeyword(groups[6]?.value as String)
    } else if (groups[7]?.value != null) {
        MalSymbol(groups[7]?.value as String)
    } else {
        throw MalReaderException("Unrecognized token: " + next)
    }
}
